### ISSUE TITLE
where can I ask questions / get help?

### ISSUE THREAD

**Issue by BrianHicks** [facebook/buck2#115]
Hey! I have a bunch of beginner-level questions. What's an appropriate place to ask them? I don't want to clutter this repo with a bunch of small questions whose answer is probably "that works like Buck 1" (which I haven't used 😅)

**Comment by ndmitchell** [comment]
Here is a good place to ask. We'll close the question once you are done, and hopefully improve the buck2 docs, so no clutter but will driven improvements in our documentation at https://buck2.build/.

**Comment by ndmitchell** [comment]
1. The rule should not look at any files that it doesn't get given from the user. To do that, you'd add an attribute `"srcs": attrs.list(attrs.source())`, and then in the `BUCK` file pass `glob(["srcs/**/*.elm"])`. If you have two types of source file (e.g. I suspect `elm.json` is either config or a project file?) you'd probably want a separate attribute for that.
2. We only have isolated commands when running through remote execution (https://buck2.build/docs/remote_execution/). In many cases, just calling `run` will require things like `elm.json` as arguments. If there is a `elm make` mode where you can pass `--project` or similar, then `["--project", ctx.attrs.project]` will add the dependency. It's better to run the tools in "explicit" mode if available. If not you can do `cmd_args(["elm", "make", "--docs", docs.as_output()]).hidden(ctx.attrs.project, ctx.attrs.srcs)` and it will know they are required for the command.
3. Yes, a toolchain is the way to go. No great docs on this unfortunately - they're all a bit complex, and you probably need something fairly simple. At it's simplest a toolchain is a rule (https://github.com/facebook/buck2/blob/main/prelude/toolchains/ocaml.bzl) that produces a provider (https://github.com/facebook/buck2/blob/f9755e80863a1b9ca4c1358575a7a317a4707919/prelude/ocaml/ocaml_toolchain_types.bzl#L12-L43), and then you add an automatic `_elm_toolchain` attribute to `elm_docs` - OCaml isn't a bad reference and is fairly simple. I'd probably go a bit further before investing in a toolchain.

**Comment by BrianHicks** [comment]
Ah, great. Thank you! `hidden` is gonna get a lot of play here—the Elm compiler does a lot of things implicitly with packages.

About isolation: is that planned-for at all? One of the reasons I'm excited by Buck (or other similar tools) is to get better enforcement around dependencies. I suppose if we eventually had a remote build server we'd get it anyway though!

Is there a specific point at which you'd recommend looking at a toolchain or it more like "you have bigger fish to fry right now"?

**Comment by BrianHicks** [comment]
a couple more miscellaneous questions:

---

Every `elm` invocation requires there to be an `elm.json` in the working directory. As you point out, it'd be nice to have it specified as `--project` or similar, but that's not an option (even as an environment variable.) I can probably have every invocation set it explicitly by requiring a `root` attr or similar, but it'd be nice to have it just be the default since you can't ever change it. Is there a way to construct a source or set a relative file (e.g. `"elm.json"`) as a default? When I try it, Buck says it can only be set to a relative file in the `BUCK` file. Fair, but a bit repetitive in this case?

---

What is the normal pattern for reading a config file for dependency information and creating calls to (say) `elm_library` based on that? Is that something I'll need to use an external tool for or can Buck 2 handle it internally?

**Comment by ndmitchell** [comment]
* For isolation, you can set up buildbarn today and have a local remote execution server which gives you isolation. See #105 for a potential idea of basically baking something like buildbarn into Buck2, so we do that for you behind the scenes if you want isolation. The alternative would be to define your own `elm.py` that takes everything explicit, creates a temp dir, changes directory and copies/symlinks everything you explicitly passed, then does the build. A bit manual perhaps, but fairly simple to do.
* For toolchain, you want that when you move to remote execution without base images (e.g. if you have `elm` in the base image, you are probably fine) or when you want to productionise and share it with other Elm projects (eventually we'd love to have these rules in the Buck2 prelude, which would require a toolchain). For when, I guess I don't think it will help you evaluate Buck2 for your use case (no hard bits there), but it will help you productionise. So when you have decided that Buck2 is the one true build system for you, I'd start thinking about a toolchain.
* For a `elm.json` not repeating you can define a wrapper `def elm_docs_helper(**kwargs): elm_docs(root = "elm.json", **kwargs)`. You can even define that in `elm.bzl` as `elm_docs` and then its like there is an implicit attribute, it just happens to get filled in at the users location.
* For dependencies, do you mean which Elm libraries it depends on? Usually you'd vendor those in and list them explicitly. Although I'm not sure that's quite what you mean. Perhaps an example?

**Comment by BrianHicks** [comment]
All very helpful, thanks again! The wrapper makes a lot of sense; I'm gonna have to do that right away!

Speaking of changing working directory, by the way, is there a nice way to do that in `cmd.actions.run` or `cmd_args` or does it require invoking bash or something? I see `relative_to`, but that only seems to change the output directories?

I've got a couple of examples for dependencies:

- You can get packages from packages.elm-lang.org. When I've done Elm packaging with Nix, there's a way to analyze the `elm.json` ([example](https://github.com/NoRedInk/noredink-ui/blob/master/elm.json)) to get the package dependencies. I'd like to do the same (or similar) in Buck so that we can get better caching.
- Elm packages and projects also specify source roots. I would hypothetically like to tell Buck to depend on all the Elm files it can glob out of those prefixes so that there's only one source of truth for build information.

**Comment by ndmitchell** [comment]
The only way to change working directory is with a shell script. For most tools (which take explicit arguments) that gives you a much higher chance of getting it right than changing directory and then finding things relatively. That said, if you do an explicit elm wrapper, I'd expect that to take everything explicitly, and then change to a temporary directory.

For packages, you can either put them on the Buck2 graph or leave them off as untracked dependencies (with the assumption they are on the base image). The two options:

1. If you put them on the graph, you usually need to vendor them in - have something that copies/pastes all their code into your project. E.g. we use the [Reindeer tool](https://github.com/facebookincubator/reindeer) to do that for Rust dependencies. A bit more work, but works seamlessly on remote execution.
2. If you put them off the graph they are being found in the base image, so you'd need to make sure they were provided by Nix or a Docker image on remote execution. Easier to get started.

For source roots, perhaps a glob in the macro layer?

**Comment by BrianHicks** [comment]
about testing:

- is there a way to tell Buck that if it has already run some tests for a specific set of inputs, it need not run them again?
- is there a way to mark contended resources? In Elm, there's a compilation artifacts directory (`elm-stuff`.) It's not *completely* safe to have multiple instances of the compiler running at once on a single one (there is some file locking stuff but there's a crucial gap that means files can get truncated sometimes.)
- is there some equivalent to Shake's `batch` in Buck? For example, I can run the formatting tool in its check mode in Shake, and it will be invoked so many times, only on the files that have changed since it was last invoked. I'd like something similar in Buck, if possible! Failing that, I'd like to split up the `elm-format` invocations, one per file, but I don't see an obvious way to do that. Is it possible?
- (I'm thinking probably this is "no" but I'll ask anyway.) Is there a way for a command to fix up a source file in Buck? Formatters are my main use case, but `elm-review` (a linter) also has automatic fixes that I feel confident enough in to always apply.

In case it's helpful, here's my `elm-format` rule:

<details><summary><code>elm-format</code> in <code>elm.bzl</code></summary>

```starlark
def _elm_format(ctx: "context"):
    command = cmd_args(["elm-format", "--validate"])
    command.add(ctx.attrs.srcs)

    if not ctx.attrs.srcs:
        fail("I need at least one Elm file to format, but got none.")

    return [
        ExternalRunnerTestInfo(
            type = "elm_format",
            command = [command],
        ),
        DefaultInfo(),
    ]

elm_format = rule(
    impl = _elm_format,
    attrs = {
        "srcs": attrs.list(attrs.source()),
    }
)
```

</details>

**Comment by krallin** [comment]
> is there a way to tell Buck that if it has already run some tests for a specific set of inputs, it need not run them again?

You can't do that I'm afaid

> is there a way to mark contended resources? In Elm, there's a compilation artifacts directory (elm-stuff.) It's not completely safe to have multiple instances of the compiler running at once on a single one (there is some file locking stuff but there's a crucial gap that means files can get truncated sometimes.)

Not currently (though that's actually being added, at least for tests). We would recommend designing your rules to avoid needing this kind of mutual exclusion however.

> (I'm thinking probably this is "no" but I'll ask anyway.) Is there a way for a command to fix up a source file in Buck? Formatters are my main use case, but elm-review (a linter) also has automatic fixes that I feel confident enough in to always apply.

This is indeed generally something we would advise against. The way you can integrate with such linters is to have them emit structured data representing the changes that need to happen (either in a subtarget or in a provider that you then access via a BXL script), and then an external tool can call Buck to obtain this data and then apply the changes once Buck is done working.

> is there some equivalent to Shake's batch in Buck? For example, I can run the formatting tool in its check mode in Shake, and it will be invoked so many times, only on the files that have changed since it was last invoked. 

If you follow the suggestion above of having the formatter's "suggested changes" expressed as the output of running an action, then that will indeed only re-run if the inputs have changed.

> I'd like something similar in Buck, if possible! Failing that, I'd like to split up the elm-format invocations, one per file, but I don't see an obvious way to do that. Is it possible?

Again, assuming you follow the advice above, this might look like (not tested):

```starlark
def _elm_library(ctx: "context"):
  if not ctx.attrs.srcs:
        fail("I need at least one Elm file, but got none.")

  suggested_changes = []
  for src in ctx.attrs.src:
    suggestion = ctx.actions.declare_output("__suggested_changes__", src.short_path)
    # Obviously making up flags here, you can write a wrapper script as needed.
    ctx.actions.run(
      ["elm-format", "--suggest-changes", src, "--out", suggestion],
      category = "elm-format", identifier = src.short_path
    )
    suggested_changes.append(suggestion)
  
    return [
         # Would be good to expose a default output of course too
        DefaultInfo(sub_targets= {"format": [DefaultInfo(suggested_changes)]}),
    ]

elm_library = rule(
    impl = _elm_library,
    attrs = {
        "srcs": attrs.list(attrs.source()),
    }
)
```

I'd strongly recommend not trying to model this formatting as a test, it's pretty far away from the use case for tests

**Comment by BrianHicks** [comment]
Ooh! Such an interesting idea. OK, I can certainly move away from formatters-as-tests. Thanks for that advice. I guess in that case, you'd have some CI job running to make sure there are no formatting changes to be made instead of using `buck2 test`?

Also I guess when I'm asking "how can I avoid re-running tests", what I'm actually asking is this: is there a way to run tests only for things whose dependencies have changed? The [docs I'm seeing](https://buck2.build/docs/rule_authors/test_execution/) imply that may be Tpx's job, though?

**Comment by BrianHicks** [comment]
Ok, neat! Thank you both again, this has been very helpful. I'm almost done moving everything in this small repo to Buck. I have a few remaining questions, though:

- How exactly can I get a list of targets associated with a file? I don't see how to do it in the target language. The workflow I'd like to enable would be something like…
  1. see what files changed in `git`
  2. find the targets for those files
  3. build/test only those (or build everything but test only those)
- Is there a way to build all matching sub-targets? I defined a `format` sub-target as @krallin suggested, and tried `buck2 build '//...[format]`. Didn't work! I'm not sure how I can do that in a BXL script either—maybe I have to do a uquery for `kind`s that I know have a `format` sub-target?
- is there a prescribed way to coordinate networked processes in buck other than "do whatever you want in a bash script"? I'm thinking a database or server for e2e tests, for instance.

**Comment by krallin** [comment]
> How exactly can I get a list of targets associated with a file? I don't see how to do it in the target language. The workflow I'd like to enable would be something like…

This should work:

```
buck2 uquery 'owner(%s)' $FILE
```

> Is there a way to build all matching sub-targets

Unfortunately not, bxl is likely what you want here, yes. In BXL you don't really need to filter by the kind, you can just do analysis then use `analysis[DefaultInfo].sub_targets.get("format")` and see whether that's `None` or not.

> is there a prescribed way to coordinate networked processes in buck other than "do whatever you want in a bash script"? I'm thinking a database or server for e2e tests, for instance.

Not currently

**Comment by BrianHicks** [comment]
when you say "do analysis"—what kind of analysis? Do you mean to just get all the targets and then look inside them to see if they have a "format" target?

**Comment by BrianHicks** [comment]
amazing, thank you. I also just managed to trigger exactly the builds/tests I need with that uquery. Such a big improvement!

Y'all are really making my day over here. ❤️
